chromium_src/net/socket/socks5_client_socket_auth.cc

```diff
@@ -0,0 +1,146 @@
+// Copyright 2017 The Brave Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "net/base/io_buffer.h"
+#include "net/socket/socks5_client_socket_auth.h"
+
+namespace net {
+
+SOCKS5ClientSocketAuth::SOCKS5ClientSocketAuth(
+    std::unique_ptr<ClientSocketHandle> transport_socket,
+    const HostResolver::RequestInfo& req_info,
+    const HostPortPair& proxy_host_port)
+    : SOCKS5ClientSocket(std::move(transport_socket), req_info),
+      proxy_host_port_(proxy_host_port),
+      next_state_(STATE_INIT_WRITE) {
+}
+
+SOCKS5ClientSocketAuth::~SOCKS5ClientSocketAuth() = default;
+
+const std::string& SOCKS5ClientSocketAuth::username() {
+  return proxy_host_port_.username();
+}
+
+const std::string& SOCKS5ClientSocketAuth::password() {
+  return proxy_host_port_.password();
+}
+
+bool SOCKS5ClientSocketAuth::do_auth() {
+  return username().size() != 0 || password().size() != 0;
+}
+
+uint8_t SOCKS5ClientSocketAuth::auth_method() {
+  if (!do_auth())
+    return 0x00;
+  return 0x02;
+}
+
+static const size_t kSOCKSAuthUsernamePasswordResponseLen = 2;
+
+int SOCKS5ClientSocketAuth::Authenticate(
+    int rv, ClientSocketHandle& transport, NetLogWithSource& net_log,
+    CompletionCallback& callback) {
+  if (!do_auth()) {
+    DCHECK_EQ(OK, rv);
+    return OK;
+  }
+  do {
+    switch (next_state_) {
```

Comment:
we don't want to copy all of this, we can call `SOCKS5ClientSocket::Authenticate` here

Reply:
I don't understand.  SOCKS5ClientSocket::Authenticate does nothing.  This function is where the actual authentication logic appears.

Reply:
That is, SOCKS5ClientSocketAuth::Authenticate is not a copy of any logic; it is the main substantive addition to the functionality in this subclass.

Reply:
wasn't the original patch much smaller?

Reply:
I thought it was added to `SOCKS5ClientSocket::DoLoop`?

Reply:
Nope, that code is basically verbatim from the original patch.  The original patch -- now https://github.com/riastradh-brave/muon/tree/riastradh-socks5-auth-orig -- factored the state machine for the authentication part into a separate class `SOCKSAuthUsernamePassword`.  The purpose was to minimize changes to the `SOCKS5ClientSocket` class itself, although I still put it into the original source file.